src/lib.rs

```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+                Either::A(futures::future::ok((id_to_name, instances)))
+            } else {
+                Either::B(futures::future::err(format_err!(
+                    "not all instances were launched within the time limit"
+                )))
+            }
+        })
+            .and_then(move |(id_to_name, instances)| {
+                // 3. wait until all instances are up
+                // note that we *don't* do this check if we have no instances, b/c empty instance list
+                // means "list all instances".
+                let mut desc_req = rusoto_ec2::DescribeInstancesRequest::default();
+                let term_instances = instances.clone();
+                desc_req.instance_ids = Some(instances);
+
+                futures::future::loop_fn(id_to_name, move |id_to_name| {
+                    ec2.describe_instances(&desc_req).map(move |reservations| {
+                        let mut machines = HashMap::new();
+                        for reservation in reservations.reservations.unwrap_or_else(Vec::new) {
+                            for instance in reservation.instances.unwrap_or_else(Vec::new) {
+                                match instance {
+                                    rusoto_ec2::Instance {
+                                        instance_id: Some(instance_id),
+                                        instance_type: Some(instance_type),
+                                        private_ip_address: Some(private_ip),
+                                        public_dns_name: Some(public_dns),
+                                        public_ip_address: Some(public_ip),
+                                        ..
+                                    } => {
+                                        let name = id_to_name[&instance_id].clone();
+                                        let machine = Machine {
+                                            ssh: None,
+                                            instance_type,
+                                            private_ip,
+                                            public_ip,
+                                            public_dns,
+                                        };
+
+                                        trace!(log, "instance ready"; "set" => &name, "ip" => &machine.public_ip);
+                                        machines.entry(name).or_insert_with(Vec::new).push(machine);
+                                    }
+                                    _ => {
+                                        return futures::future::Loop::Continue(id_to_name);
+                                    }
                                 }
-                                Some(e) => Either::B(futures::future::err(e)),
                             }
+                        }
+
+                        futures::future::Loop::Break(machines)
+                    })
+                }).map(|machine| (machine, term_instances))
```

Comment:
Hmm, I don't think this will work? `WaitForInstancesResult::NotOk` would be turned into an `Err` around line 828, which would mean that any instances contained within it would not be mapped up here, which means that they wouldn't be terminated!

Reply:
I'm not sure how to resolve this. I'm actually not even sure how it works in the original codebase. I thought I just inlined a closure, but it seems like that changed the behavior somehow?